app/src/msg.c

```diff
@@ -0,0 +1,78 @@
+/*******************************************************************************
+*   (c) 2018 - 2022 Zondax AG
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+********************************************************************************/
+
+#include <stdio.h>
+#include "coin.h"
+#include "zxerror.h"
+#include "zxmacros.h"
+#include "zxformat.h"
+#include "app_mode.h"
+#include "crypto.h"
+#include "parser_utils.h"
+#include "crypto_helper.h"
+#include "tx.h"
+
+#define IS_PRINTABLE(c) ((c >= 0x20 && c <= 0x7e) || (c >= 0x80 && c <= 0xFF))
+
+zxerr_t msg_getNumItems(uint8_t *num_items) {
+    zemu_log_stack("msg_getNumItems");
+    *num_items = 1;
+    return zxerr_ok;
+}
+
+zxerr_t msg_getItem(int8_t displayIdx,
+                     char *outKey, uint16_t outKeyLen,
+                     char *outVal, uint16_t outValLen,
+                     uint8_t pageIdx, uint8_t *pageCount) {
+    ZEMU_LOGF(200, "[msg_getItem] %d/%d\n", displayIdx, pageIdx)
+   
+    const uint8_t *message = tx_get_buffer();
+    const uint16_t messageLength = tx_get_buffer_length();
+
+     switch (displayIdx) {
+         case 0: {
+            //Non Printable Chars Counter
+            bool npc = false;                                
+            for (uint32_t i=0; i < messageLength; i++) {
+                if(!IS_PRINTABLE(message[i])) {
+                    npc=true;
```

Comment:
previous version asked for at least 40% of npc to display the message as binary/hex